src/vsphere_cpi/lib/cloud/vsphere/selection_pipeline.rb

```diff
@@ -0,0 +1,125 @@
+module VSphereCloud
+  class SelectionPipeline
+    include Enumerable
+    include Logger
+
+    module StableRandom
+      def self.[](object)
+        @memo ||= {}
+        @memo[object] ||= Random.rand
+      end
+    end
+
+    # @return [Object] the object to select placements for
+    attr_reader :object
+
+    def self.with_filter(*args, &block)
+      filter_list << { args: args, block: block }
+    end
+
+    def self.filter_list
+      @filter_list ||= []
+    end
+
+    def self.with_scorer(*args, &block)
+      scorer_list << { args: args, block: block }
+    end
+
+    def self.scorer_list
+      @scorer_list ||= []
+    end
+
+    # @param object [Object] the object to select placements for
+    def initialize(object, *args)
+      raise ArgumentError, 'No gather block provided' unless block_given?
+
+      @object = object
+      @gather = Proc.new
+
+      self.class.filter_list.each do |filter|
+        with_filter(*filter[:args], &filter[:block])
+      end
+
+      self.class.scorer_list.each do |scorer|
+        with_scorer(*scorer[:args], &scorer[:block])
+      end
+    end
+
+    def inspect
+      "#<SelectionPipeline @filter_list #{filter_list} @scorer_list #{scorer_list}>"
+    end
+
+    def accept?(placement)
+      result = filter_list.all? do |filter|
+        result = filter.call(placement, @object)
+        logger.info("Rejecting #{placement.inspect}") unless result
+        result
+      end
+      logger.info("Selecting (All filter passed) : #{placement.inspect}") if result
+      result
+    end
+
+    def each
+      return enum_for(:each) unless block_given?
+
+      logger.info("Initiating #{inspect}")
+
+      gather.select do |placement|
+        logger.info("Applying filters to #{placement.inspect_before}")
+        accept?(placement)
+      end.sort(&compare_placements).reverse.each(&Proc.new)
+    end
+
+    def with_filter(*args)
+      if block_given?
+        unless args.empty?
+          raise ArgumentError, "wrong number of arguments (given #{args.length}, expected 0)"
+        end
+        filter_list << Proc.new
+      else
+        raise ArgumentError, "0 arguments passed, expected atleast 1 arg or a block if no args provided" if args.empty?
+        filter_list.concat(args)
+      end
+      self
+    end
+
+    def with_scorer(*args)
+      if block_given?
+        unless args.empty?
+          raise ArgumentError, "wrong number of arguments (given #{args.length}, expected 0)"
+        end
+        scorer_list << Proc.new
+      else
+        raise ArgumentError, "0 arguments passed, expected at least 1 arg or a block if no args provided" if args.empty?
+        scorer_list.concat(args)
+      end
+      self
+    end
+
+    private
+
+    def compare_placements
```

Comment:
While `#compare_placements` conforms to Ruby's `Comparable` module's semantics I really dislike this use in the selection pipeline because it's unclear which placement is considered "better" is not obvious.

I think each scorer should be passed `p1` and `p2` and return the "better" placement or `nil` if the two placements are considered equivalent.

Reply:
I do not understand the intent completely. Can this be taken up in a separate story to work on scorers?
I want to design them in a way to add some logging too.

Reply:
The problem is that unfortunately some of the code in `VmPlacementSelectionPipeline` is just wrong:

```ruby
    with_scorer do |p1, p2|
      p1.migration_size <=> p2.migration_size
    end

    with_scorer do |p1, p2|
      p1.free_memory <=> p2.free_memory
    end
```

Notice that `migration_size` and `free_memory` are both used to determine the scoring for `VmPlacement`s. However these two metrics have an inverse relationship: **lower** `migration_size` is good while **higher** `free_memory` is good.

This means that either:

1. The `migration_size` scorer is wrong here and needs to be:

   ```ruby
   p2.migration_size <=> p1.migration_size
   ```

2. Or the `free_memory` scorer is wrong here and needs to be:

   ```ruby
   p2.free_memory <=> p1.free_memory
   ```

Unfortunately I can't tell which one is wrong which is why I think that we need to rethink `compare_placements`.

That being said we can merge this and fix later if you want to avoid blocking the other tracks of work.

Reply:
So,  

`p1.migration_size <=> p2.migration_size` sorts in ascending order but 

`-(p1.migration_size <=> p2.migration_size)` should sort in descending order.

Having a negative sign should make it easier to write each scorer. 

But I agree that when to use negative and when not to use negative is not very intuitive while writing and can lead to bugs. 

For now it seems to be a quick and good fix. Thoughts?

Reply:
this bug should have resulted in a test failure right?

Reply:
just saw the test changes

Reply:
yeah, it should have failed. Test was wrong. It expected everything to be in descending order.
Now it is corrected.